Neural networks improvements

dmonllao opened the pull request:
Hi, I've tried the current MultilayerPerceptron with a sparse dataset I have and it greatly outperformed linear classifiers like perceptron and adaline. I would like to give some love to php-ml neural nets and make them usable for production (partial training, persistence...) I am keen to do it even if it is not perfect or if in future it is moved to a separate project under php-ai. This patch would be the first of a serie followed by partial training and persistence. I am keen to update the docs page as well if we go with these changes.

Some comments:

- I am not sure what is the purpose of Backpropagation's $desiredError argument. Well, I can understand the purpose what I am not sure about is if it is a good idea, isn't it encouraging people to overfit the model to the training set? I am proposing to follow linear classifier models, we keep the train($samples, $targets) signature and we move $maxIterations param to the constructor, that would make much more sense having both train and partialTrain methods. MLP is a good algorithms for batch training.

-  Shouldn't we only consider the prev layer (or next depending on the POV :P) sigmas when updating neuron's weights? It seems that we are currently using all further layers's sigmas. I have an example of how nn with multiple hidden layers would not work with the current code https://github.com/dmonllao/php-ml/compare/nn-failures~...nn-failures (testBackpropagationForXORLearningMultiLayer) The PR branch contains a fix for this

- It seems that the current tests are passing because the process stops when $desiredError is reached, if the process continues it randomly fails. You can see what I mean in https://github.com/dmonllao/php-ml/compare/nn-failures~...nn-failures (testBackpropagationForXORLearningNoDesiredError). I think (I haven't verified it, just an idea) that the current test is tricky for a 2-2-1 neural network because depending on the random initialization getSigmaForNeuron just sums up all weights * sigmas. Also 40000 iterations seems far more than it should be needed for a case like this, we may be getting passes just because we iterate lots of times. I've updated the conflictive test to make it [1, x] = 'a' and [x, 1] = 'b'. I've tried the same scheme in MLPClassifier and it also failed. This is probably related to #87 and we should be able to close it with these changes.

- I've simplified the targets format and updated it according to other algorithms (single array). Internally the output layer contains 1 node for each different class but this is transparent for the user, who uses the same format than in other php-ml algorithms. If we are interested in allowing MLP usable without specifying the list of classes we can still do it in train() once they support both train() (reset all previous stuff including getting rid of the last layer as the training dataset may include different targets) and partialTrain

-  Following this last point, I've added predict() method to the MLP which I guess most of the people interested in using php nns missed. It just feed-forwards each sample input and gets the output for each of the output layer neurons, the returned prediction is the one with an higher score

- I've added support for non-numeric targets

- In order to simplify the public API of php-ml's NN I've been very tempted to move Backpropagation initialization to MLP's constructor and add Backpropagation's params to MLP signature to forward them to Backpropagation. This way MLP (along with this PR changes) could implement Estimator interface. Ideally people would not need to know about the algorithm internals. I would like to hear your thoughts about it, I could do it as part of this PR.

dmonllao commented:
I didn't run all tests, I will send a patch updating them in a few mins

EvulMastah commented:
This also fixes https://github.com/php-ai/php-ml/issues/87 when accepted. I have waited more than 150 iterations, whereas it used to fail always under 100.

itcraftsmanpl commented on src/Phpml/NeuralNetwork/Network/MultilayerPerceptron.php:
```diff
@@ -11,28 +11,78 @@
+     */
+    public function getTargetClass($target): int
+    {
+        if (in_array($target, $this->classes) === false) {
```
`if(!in_array())`

itcraftsmanpl commented:
First: @dmonllao grate work. You motivate me to find more time to develop php-ml :D :+1: :+1: 

This is why i introduce desire error:
```
for ($i = 0; $i < $this->maxIterations; ++$i) {
    $this->trainSamples($samples, $targets);
}
```
Sometimes there is no need to iterate so many times. The NN topic is new for me and this classes was prof of concept so i'm not expert, but i'm able to merge this.

EvulMastah commented on src/Phpml/NeuralNetwork/Network/MultilayerPerceptron.php:
```diff
@@ -11,28 +11,78 @@
+    public function __construct(int $inputLayerFeatures, array $hiddenLayers, array $classes, ActivationFunction $activationFunction = null)
     {
-        if (count($layers) < 2) {
+        if (count($hiddenLayers) < 1) {
```
`if (empty($hiddenLayers))` would be better, as you expect at least 1 element.

EvulMastah commented on src/Phpml/NeuralNetwork/Network/MultilayerPerceptron.php:
```diff
@@ -11,28 +11,78 @@
+                $max = $value;
+            }
+        }
+        return $this->classes[$predictedClass];
```
Is `$predictedClass` *guaranteed* to be found? Otherwise you will be accessing `null` index, which is undefined.

dmonllao replied:
I can't think of any remote case where this could happen, classes are set in the constructor and getOutput returns a class for each output layer node.

EvulMastah commented on src/Phpml/NeuralNetwork/Training/Backpropagation.php:
```diff
@@ -22,99 +22,104 @@ class Backpropagation implements Training
+                $value = 1;
+            } else {
+                $value = 0;
+            }
```
Branching could be reduced with:
```php
$value = 0;
if ($targetClass == $key) {
    $value = 1;
}
```

itcraftsmanpl replied:
or even with `$value = (int) $targetClass == $key`

itcraftsmanpl replied:
also could we use `===`?

dmonllao commented:
The conflict should be fixed now, I've just rebased on top of master.

dmonllao commented:
- Thanks for your comments. I've updated the patch according to them, but are those coding style rules specified somewhere or are they just your personal preferences?
- I got not reply about MLP initialising Backpropagate so I've done it :P now the public interface of the multilayer perceptron is even simplier (I've updated docs as well)
- I've abstracted MultilayerPerceptron stuff, created a new MLPClassifier and merged backpropagation tests into MLPClassifier one. Related with this, I've deleted MLPRegressor, it was not documented and had no tests I doubt if it was working properly before. I am keen to add it back at some point but I need to read more nn stuff to be confident enough to change some parts of the code. A part of the backpropagation process will probably have to be abstracted as well to include a MLP regressor. Same with MultilayerPerceptron constructor, MLPRegressor may have to force 1 single output unit
- Docs updated as well

itcraftsmanpl commented:
For coding styles you can use newest version of php-cs-fixer. Maybe we can put `MLPClassifier` in `Classification` category (readme) or better leave it in NN?

dmonllao commented:
yeah, I added it to classification, following what you did with other models

dmonllao commented:
Just woke up and first thing I see https://travis-ci.org/dmonllao/php-ml/jobs/233187669 grrr, anyway, I will continue pushing changes during these days, sorry for the inconvenience

itcraftsmanpl commented:
Is harder than we think ...
